core/src/main/java/org/opensearch/sql/executor/pagination/PaginatedPlanCache.java

```diff
@@ -8,152 +8,160 @@
 import org.opensearch.sql.planner.physical.PaginateOperator;
 import org.opensearch.sql.planner.physical.PhysicalPlan;
-import org.opensearch.sql.planner.physical.ProjectOperator;
 import org.opensearch.sql.storage.StorageEngine;
-import org.opensearch.sql.storage.TableScanOperator;
 
 /**
  * This class is entry point to paged requests. It is responsible to cursor serialization
  * and deserialization.
  */
-@RequiredArgsConstructor
-public class PaginatedPlanCache {
+public class PaginatedPlanCache implements AutoCloseable {
   public static final String CURSOR_PREFIX = "n:";
-  private final StorageEngine storageEngine;
+
+  /**
+   * See {@link SerializationContext#engine}.
+   */
+  public PaginatedPlanCache(StorageEngine engine) {
+    SerializationContext.engine = engine;
+  }
 
   public boolean canConvertToCursor(UnresolvedPlan plan) {
     return plan.accept(new CanPaginateVisitor(), null);
   }
 
   /**
-   * Converts a physical plan tree to a cursor. May cache plan related data somewhere.
+   * Actually this happens only on OpenSearch node shutdown, because lifetime of PaginatedPlanCache
+   * is equals to SQL plugin lifetime.
    */
-  public Cursor convertToCursor(PhysicalPlan plan) throws IOException {
-    if (plan instanceof PaginateOperator) {
-      var cursor = plan.toCursor();
-      if (cursor == null) {
-        return Cursor.None;
-      }
-      var raw = CURSOR_PREFIX + compress(cursor);
-      return new Cursor(raw.getBytes());
-    }
-    return Cursor.None;
+  @Override
+  public void close() throws Exception {
+    SerializationContext.engine = null;
   }
 
   /**
-   * Compress serialized query plan.
-   * @param str string representing a query plan
-   * @return str compressed with gzip.
+   * An auxiliary class, which provides an entry point for serialization and deserialization of
+   * the plan tree. It doesn't serialize itself, it calls {@link SerializablePlan#writeExternal}
+   * of the given plan. For deserialization, it loads a {@link SerializablePlan.PlanLoader} and
+   * invokes it.
    */
-  String compress(String str) throws IOException {
-    if (str == null || str.length() == 0) {
-      return "";
+  @Data
+  public static class SerializationContext implements Externalizable {
+    private PaginateOperator plan;
+    // TODO get engine from GUICE if possible
+    /**
+     * We have to make {@link PaginatedPlanCache.SerializationContext} a non-static class, we can
+     * avoid setting engine statically, but in that case deserialization fails with exception
+     * 'no default (no-arg) constructor', even if it has one. This is caused because default
+     * constructor ofr a non-static class isn't available from outside (from serialization engine).
+     * The engine is not being serialized, but it is required for deserialization.
+     * See usages of {@link SerializablePlan.PlanLoader}, especially in `OpenSearchPagedIndexScan`.
+     */
+    private static StorageEngine engine;
```

Comment:
I know you have a different version of this code.

IIRC this class is necessary to pass an instance of StorageEngine to `OpenSearchPagedIndexScan` loader, right? Have you looked into how to use Guice to do that?

I know you could do it with [`ObjectInputStream.resolveObject`](https://docs.oracle.com/javase/7/docs/api/java/io/ObjectInputStream.html#resolveObject(java.lang.Object)). I'm looking into how to do it with Guice -- I expect it will be cleaner.

Reply:
Trying to update `OpenSearchPluginModule` is my next step

Reply:
Fixed (I hope) in 490af96.